Fix: Support pageAction show/hide wrapping

rpl opened the pull request:
This PR contains the changes to the api-metadata.json from #9 and it also applies the changes needed into the polyfill wrappers to be able to customize the wrapper behavior for the API methods that returns a Promise on Firefox but do not accept a callback on Chrome (which is the case for pageAction.show and hide).

rpl commented:
@Rob--W do you mind to take a look at this PR?

similarly to #57 it is going to change the api-metadata.json but it also needs some additional changes in the polyfill wrappers, to be able to customize the behavior for the kind on chrome incompatibility that pageAction.show/hide presents.

Rob--W commented:
This doesn't look like a forward-compatible change to me.

Yes, it would fix the current error.
However, if Chrome were to support the callback parameter (which I expect to happen - https://github.com/mozilla/webextension-polyfill/pull/13#issuecomment-269387289, I've wanted to submit a patch myself, but that's near the bottom of my priority list), then the polyfill will cause an error again.

How about trying to call with callback, and if a synchronous (schema) error about unsupported callbacks occurs, invoke the method without callback?

```javascript
try { chrome.browserAction.setTitle(1, function(){}); } catch (e) { console.log(e.message); }
```

gives the following result (tested Chrome 34 - 61:

> Invocation of form browserAction.setTitle(integer, function) doesn't match definition browserAction.setTitle(object details)

and when `--native-crx-bindings=1` is set (tested with Chrome 61) (Chrome is working on [replacing JS-implemented extension bindings with C++](https://bugs.chromium.org/p/chromium/issues/detail?id=653596)):

> Error in invocation of browserAction.setTitle(object details): Too many arguments.


PS. Why does browserAction.setTitle not appear in api-metadata.json?

Rob--W commented on src/browser-polyfill.js:
```diff
@@ -153,6 +150,27 @@ if (typeof browser === "undefined") {
+              console.warn(`${name} API method doesn't seem to support the callback parameter, ` +
+                           "falling back to call it without a callback: ", cbError);
+
+              // Update the API method metadata, so that the next API call will not try to
```
Move this after line 170, to make sure that the API call is actually valid before considering the API callback-less.

rpl replied:
:+1: great call!

Rob--W commented:
The `browserAction` API map is missing entries like `browser.runtime.enable`.

I guess that the JSON file is manually edited. Maybe it helps if you write a quick script to load the JSON files from mozilla-central and compare it with the existing definitions to see if anything is missing.

rpl commented:
@Rob--W I've updated this PR based on your last round of review comments and updated the test cases accordingly, let me know how it looks to you.

Thanks again for you help!

Rob--W commented on src/browser-polyfill.js:
```diff
@@ -132,7 +132,48 @@ if (typeof browser === "undefined") {
+          function callAPIWithNoCallback() {
+            try {
+              target[name](...args);
+              if (chrome.runtime.lastError) {
```
This `chrome.runtime.lastError` check seems misplaced. The `target[name](...args)` call is asynchronous, so the error will never be set to that.

Worse, with the current code, if someone tries to use call e.g. `browser.pageAction.show` inside a function where `lastError` is set, then the promise will always be rejected.

rpl replied:
:+1: it was already supposed to be removed and I forgot, thanks for pointing it out.

Rob--W commented on src/browser-polyfill.js:
```diff
@@ -132,7 +132,48 @@ if (typeof browser === "undefined") {
+            try {
+              target[name](...args, makeCallback({resolve, reject}, metadata));
+            } catch (cbError) {
+              console.warn(`${name} API method doesn't seem to support the callback parameter, ` +
```
Under normal circumstances, this message will be displayed once.

This message will be displayed more than once if the extension passes the wrong arguments. Do you consider that to be OK? (personally I think that it is fine, since the code is simpler in this way; if you want to force the message to be displayed once, then it needs to be moved inside the `else` block above.

Why do we need log spam in the first place?

rpl replied:
I thought to emit at least a warning log mostly because it is not completely impossible that this runtime discovery can be fooled by some chrome behaviors (maybe a future chrome behavior) that the polyfill is not expecting, and it could completely hide the real error.

Emitting the warning every time we are actually falling back between the "with callback" to the "no callback" mode is fine from my current point of view, in the common scenarios it should only be produced once, and if it is produced more than once, then it could be a signal of something different between the behavior that it expect and the actual one.

Rob--W commented on src/browser-polyfill.js:
```diff
@@ -132,7 +132,48 @@ if (typeof browser === "undefined") {
+              target[name](...args);
+              if (chrome.runtime.lastError) {
+                // NOTE: Ideally this should rejects if there are runtime issues with
+                // the API call on Chrome, unfortunately currently when pageAction.show/hide
```
This whole comment block will go probably go away because of my above comment. But if you decide to re-instate the comments somewhere, generalize the affected APIs, e.g. "APIs such as pageAction.show" because pageAction.show/hide are not the only affected APIs.

rpl commented:
@Rob--W suggested changes applied in the updated patch

Rob--W commented on src/browser-polyfill.js:
```diff
@@ -132,7 +132,39 @@ if (typeof browser === "undefined") {
+              // Catch API parameters validation errors and rejects the promise.
+              reject(error);
+            }
+          }
```
Now the `callAPIWithNoCallback` logic is very simple, it can be inlined below, without try-catch (because you are in the closure of a Promise).

The second `callAPIWithNoCallback` call does not require the metadata to be updated, so it can simply be:

```javascript
target[name](...args);
resolve();
```

Treora commented:
Would be great if this can be merged indeed, I stumbled upon #24 as well. Is this PR still held back by anything?

I notice one unprocessed review [comment](https://github.com/mozilla/webextension-polyfill/pull/59#discussion_r142001949); I made a [PR](https://github.com/rpl/webextension-polyfill/pull/1) to this branch that inlines the function as indicated.

rpl commented:
@Treora Thanks a lot for your PR! I've rebased this branch, added your changes on top of the commit already part of this PR and added pageAction.show to the smoke tests that runs on Chrome.

I'm going to land this PR as soon as Travis job has been completed to run.
